### PR TITLE
Fix compatibility issue in commands

### PR DESCRIPTION
**Pull request by xisui-MSFT**
## This change addresses item #3446

Now these commands accept both folder and {folder, targetName} object.

### CODE REVIEW

**Comment by gcampbell-msft** [comment]
@xisui-MSFT Thanks for the suggestion. However, I think we should simply correct what happened in the original PR. Rather than using an object, it should be two separate parameters for the methods. 

For example:

`launchTargetPath(folder?: vscode.WorkspaceFolder | string, targetName?: string)`

This means that you can still pass them as regular parameters and it keeps the code simple. It does slightly affect the original implementatition, but the structure with the first parameter being `folder` had been there much longer and thus might break more people than the three months that it's been an object. 

Could you make those changes?

**Comment by xisui-MSFT** [comment]
@gcampbell-msft I don't think regular parameters are supported in `args` field in `inputs`, as proposed in the original PR. When `args` provided is an array, it's passed to the command as a single array argument; and when it's provided as an object, it's passed to the command as a single object argument. I think this is why the original PR uses object instead of normal parameters.

**Comment by gcampbell-msft** [comment]
> @gcampbell-msft I don't think regular parameters are supported in `args` field in `inputs`, as proposed in the original PR. When `args` provided is an array, it's passed to the command as a single array argument; and when it's provided as an object, it's passed to the command as a single object argument. I think this is why the original PR uses object instead of normal parameters.

Ah, I just tested and you are correct. I'll review the current state of the PR now.

File: src/extension.ts [hunk 1]
```diff
@@ -66,6 +66,7 @@ export let extensionManager: ExtensionManager | null = null;
 
 type RunCMakeCommand = (project: CMakeProject) => Thenable<any>;
 type QueryCMakeProject = (project: CMakeProject) => Thenable<string | string[] | null>;
+type FolderTargetNameArgsType = vscode.WorkspaceFolder | string | { folder?: vscode.WorkspaceFolder | string; targetName?: string };
```
Comment (gcampbell-msft) [inline src/extension.ts:69]:
```suggestion
type FolderTargetNamedArgsType = vscode.WorkspaceFolder | string | { folder?: vscode.WorkspaceFolder | string; targetName?: string };
```

File: src/util.ts [hunk 2]
```diff
@@ -738,7 +738,7 @@ export function isNullOrUndefined(x?: any): x is null | undefined {
     return (x === null || x === undefined);
 }
 
-export function isWorkspaceFolder(x?: any): boolean {
+export function isWorkspaceFolder(x?: any): x is vscode.WorkspaceFolder {
```
Comment (gcampbell-msft) [inline src/util.ts:741]: Instead of this, should we leave it as boolean and simply add a check in the method itself for `x is vscode.WorkspaceFolder`?

Comment (gcampbell-msft) [reply src/util.ts:741]: If this is possible I think I'd rather keep the generic `boolean`

**Review by gcampbell-msft — APPROVED** [review]
Approved but with a couple of changes that I think would be nice, but aren't required before merging.

File: src/util.ts (hunk 2 above)
Comment (xisui-MSFT) [reply src/util.ts:741]: It's returning a boolean. The difference is when `x is vscode.WorkspaceFolder` is used as return type, it lets the interpreter know its type when it returns true. I.E, when `x is vscode.WorkspaceFolder` is returned:
```
if (isWorkspaceFolder(x)) {
    // x is now a WorkspaceFolder
}
```

When returning boolean,
```
if (isWorkspaceFolder(x)) {
    // x is still any
}
```

File: src/extension.ts (hunk 1 above)
Comment (xisui-MSFT) [reply src/extension.ts:69]: It means this type contains `Folder` and `TargetName`. It's not meant to be an adjective. But I'm open to better naming.

File: src/util.ts (hunk 2 above)
Comment (gcampbell-msft) [reply src/util.ts:741]: Ah, that sounds great! Thanks for the explanation.

File: src/extension.ts (hunk 1 above)
Comment (gcampbell-msft) [reply src/extension.ts:69]: Ahh, that seems reasonable to me, I just read it wrong.

**Pull request merged**